src/plugin.c

```diff
@@ -90,17 +149,71 @@ EXPORT m64p_error CALL PluginGetVersion(m64p_plugin_type *PluginType, int *Plugi
-      controller[i].control = ControlInfo.Controls + i;
-
       // init controller
-      controller[i].control->Present = 1;
+      controller[i].control = ControlInfo.Controls + i;
       controller[i].control->Plugin = PLUGIN_NONE;
+
+      m64p_handle pConfig;
+
+      /* Open the configuration section for this controller */
+      char SectionName[32];
+      sprintf(SectionName, "Input-Bot-Control%i", i); 
+      if (ConfigOpenSection(SectionName, &pConfig) != M64ERR_SUCCESS) {
+          DebugMessage(M64MSG_ERROR, "Couldn't open config section '%s'", SectionName);
+          continue;
+      }
+      DebugMessage(M64MSG_INFO, "Reading %s", SectionName);
+
+      /* throw warnings if 'plugged' is missing or can't be parsed */
+      char plugged_str[256];
+      if (ConfigGetParameter(pConfig, "plugged", M64TYPE_STRING, plugged_str, 256) == M64ERR_SUCCESS) {
+          int result = -1;
+          if (sscanf(plugged_str, "%d", &result) == 1) {
+            controller[i].control->Present = result;
+            DebugMessage(M64MSG_INFO, "    plugged=%d", controller[i].control->Present);
+          } else {
+            DebugMessage(M64MSG_WARNING, "Parsing error in 'plugged' parameter for controller %d; expected an integer, but got %s", i, plugged_str);
+          }
+      } else {
+        if(i == 0) // Maintain previous behavior of always assuming the first controller is connected
+          controller[i].control->Present = 1;
+        else
+          controller[i].control->Present = 0;
+        DebugMessage(M64MSG_WARNING, "Missing 'plugged' parameter. Set to %d", controller[i].control->Present);
+      }
+
+      if (controller[i].control->Present) {
+        /* throw warnings if 'host' is missing */
+        char host_str[256];
+        memset(controller[i].host, '\0', sizeof(controller[i].host));
+        if (ConfigGetParameter(pConfig, "host", M64TYPE_STRING, host_str, 256) == M64ERR_SUCCESS) {
+            strcpy(controller[i].host, host_str);
+            DebugMessage(M64MSG_INFO, "    host=%s", controller[i].host);
+        } else {
+            strcpy(controller[i].host, DEFAULT_HOST);
+            DebugMessage(M64MSG_WARNING, "Missing 'host' parameter. Setting to %s", controller[i].host);
+        }
+
+        /* throw warnings if 'port' is missing or can't be parsed */
+        char port_str[256];
+        if (ConfigGetParameter(pConfig, "port", M64TYPE_STRING, port_str, 256) == M64ERR_SUCCESS) {
+            if(sscanf(port_str, "%d", &controller[i].port) == 1)
+              DebugMessage(M64MSG_INFO, "    port=%d", controller[i].port);
+            else
+              DebugMessage(M64MSG_WARNING, "Parsing error in 'port' parameter for controller %d; expected an integer, but got %s", i, port_str);
+        } else {
+            controller[i].port = DEFAULT_PORT;
+            DebugMessage(M64MSG_WARNING, "missing 'port' parameter. Set to %d", controller[i].port);
+        }
+      }
```

Comment:
am I right in understanding that all of this code was essentially just brought back from the normal input plugin?

Reply:
More or less. I borrowed and adapted it from there. I had trouble using the different types (e.g. M64PTYPE_BOOL) and their respective functions. Worked fine if I edited the config and set the value directly (`plugged=1`), but setting via the CLI with `Input-Bot-Control0[plugged]=1` resulted in the config file having `plugged="1"` which failed to parse as a boolean. I may be missing something, but reading the values as a string and parsing them myself worked consistently.
Weird that the `--set` puts the quotes around the value, which is probably the real problem, and not the parsing functions.
Would probably be worth more investigation and cleaning up, but I ran out of weekend and didn't want to wait on the PR 😉

Reply:
> but I ran out of weekend

:laughing: